File: packages/shared-ts/src/secret-masking.ts
const KNOWN_PREFIXES: string[] = [
  'ghp_',
  'gho_',
  'ghu_',
  'ghs_',
  'ghr_',
  'github_pat_',
  'sk-',
  'pk-',
  'xoxb-',
  'xoxp-',
  'xoxa-',
  'xoxs-',
  'AKIA',
  'ABIA',
  'ACCA',
  'ASIA',
  'sk-ant-',
  'Bearer ',
];

const SORTED_PREFIXES = [...KNOWN_PREFIXES].sort((a, b) => b.length - a.length);

function detectPrefix(value: string): string {
  for (const p of SORTED_PREFIXES) {
    if (value.startsWith(p)) return p;
  }
  return '';
}

async function hmacSha256(key: string, message: string): Promise<Uint8Array> {
  const enc = new TextEncoder();
  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    enc.encode(key),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );
  const sig = await crypto.subtle.sign('HMAC', cryptoKey, enc.encode(message));
  return new Uint8Array(sig);
}

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

type SubtleData = ArrayBufferView<ArrayBuffer>;

export async function hmacSha256Hex(key: string, message: Uint8Array): Promise<string> {
  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode(key),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );
  const sig = await crypto.subtle.sign('HMAC', cryptoKey, message as SubtleData);
  return toHex(new Uint8Array(sig));
}

export async function mask(
  sessionId: string,
  secretName: string,
  realValue: string
): Promise<string> {
  const prefix = detectPrefix(realValue);
  const remainder = realValue.slice(prefix.length);

  const hmac = await hmacSha256(sessionId + secretName, realValue);
  let hex = toHex(hmac);

  while (hex.length < remainder.length) hex += hex;
  const maskedRemainder = hex.slice(0, remainder.length);

  return prefix + maskedRemainder;
}

export interface SecretPair {
  realValue: string;
  maskedValue: string;
}

export const MIN_MASKABLE_SECRET_LENGTH = 9;

export function buildScrubber(secrets: SecretPair[]): (text: string) => string {
  const eligible = secrets.filter((s) => s.realValue.length >= MIN_MASKABLE_SECRET_LENGTH);
  if (eligible.length === 0) return (t) => t;

  const sorted = [...eligible].sort((a, b) => b.realValue.length - a.realValue.length);

  return (text: string): string => {
    let result = text;
    for (const { realValue, maskedValue } of sorted) {
      if (result.includes(realValue)) {
        result = result.split(realValue).join(maskedValue);
      }
    }
    return result;
  };
}

export function domainMatches(pattern: string, hostname: string): boolean {
  const p = pattern.toLowerCase();
  const h = hostname.toLowerCase();

  if (p === '*') return true;

  if (!p.startsWith('*.')) {
    return p === h;
  }

  const suffix = p.slice(1);

  return h.length > suffix.length && h.endsWith(suffix);
}

export function isAllowedDomain(patterns: string[], hostname: string): boolean {
  return patterns.some((p) => domainMatches(p, hostname));
}

export function matchesDomains(hostname: string, patterns: string[]): boolean {
  return isAllowedDomain(patterns, hostname);
}

File: packages/swift-server/Sources/Keychain/SecretMasking.swift
import CommonCrypto
import Foundation



private let knownPrefixes: [String] = [
    "github_pat_",
    "sk-ant-",
    "Bearer ",
    "ghp_",
    "gho_",
    "ghu_",
    "ghs_",
    "ghr_",
    "xoxb-",
    "xoxp-",
    "xoxa-",
    "xoxs-",
    "sk-",
    "pk-",
    "AKIA",
    "ABIA",
    "ACCA",
    "ASIA",
]


private let sortedPrefixes = knownPrefixes.sorted { $0.count > $1.count }

private func detectPrefix(_ value: String) -> String {
    for p in sortedPrefixes {
        if value.hasPrefix(p) { return p }
    }
    return ""
}



private func hmacSHA256(key: String, message: String) -> [UInt8] {
    let keyData = Array(key.utf8)
    let messageData = Array(message.utf8)
    var result = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    CCHmac(
        CCHmacAlgorithm(kCCHmacAlgSHA256),
        keyData, keyData.count,
        messageData, messageData.count,
        &result
    )
    return result
}

private func toHex(_ bytes: [UInt8]) -> String {
    bytes.map { String(format: "%02x", $0) }.joined()
}






public func hmacSHA256Hex(key: String, message: [UInt8]) -> String {
    let keyData = Array(key.utf8)
    var result = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    CCHmac(
        CCHmacAlgorithm(kCCHmacAlgSHA256),
        keyData, keyData.count,
        message, message.count,
        &result
    )
    return toHex(result)
}














public func mask(sessionId: String, secretName: String, realValue: String) -> String {
    let prefix = detectPrefix(realValue)
    
    
    
    
    let prefixUTF16Length = prefix.utf16.count
    let remainderUTF16Length = realValue.utf16.count - prefixUTF16Length

    let hmac = hmacSHA256(key: sessionId + secretName, message: realValue)
    var hex = toHex(hmac)

    
    while hex.count < remainderUTF16Length { hex += hex }
    let maskedRemainder = String(hex.prefix(remainderUTF16Length))

    return prefix + maskedRemainder
}


public struct SecretPair {
    public let realValue: String
    public let maskedValue: String

    public init(realValue: String, maskedValue: String) {
        self.realValue = realValue
        self.maskedValue = maskedValue
    }
}









public let minMaskableSecretLength: Int = 9










public func buildScrubber(secrets: [SecretPair]) -> @Sendable (String) -> String {
    let eligible = secrets.filter { $0.realValue.utf16.count >= minMaskableSecretLength }
    guard !eligible.isEmpty else { return { $0 } }

    let sorted = eligible.sorted { $0.realValue.count > $1.realValue.count }

    return { text in
        var result = text
        for pair in sorted {
            result = result.replacingOccurrences(of: pair.realValue, with: pair.maskedValue)
        }
        return result
    }
}







public func domainMatches(pattern: String, hostname: String) -> Bool {
    let p = pattern.lowercased()
    let h = hostname.lowercased()

    
    if p == "*" { return true }

    guard p.hasPrefix("*.") else {
        return p == h
    }

    
    let suffix = String(p.dropFirst(1))  
    return h.count > suffix.count && h.hasSuffix(suffix)
}


public func isAllowedDomain(patterns: [String], hostname: String) -> Bool {
    patterns.contains { domainMatches(pattern: $0, hostname: hostname) }
}
